src/taipy/core/data/mongo.py

```diff
@@ -0,0 +1,225 @@
+
+        super().__init__(
+            config_id,
+            scope,
+            id,
+            name,
+            parent_id,
+            last_edit_date,
+            job_ids,
+            validity_period,
+            edit_in_progress,
+            **properties,
+        )
+        self._decoder = self._properties.get(self._DECODER_KEY, DefaultJSONDecoder)
+        self._encoder = self._properties.get(self._ENCODER_KEY, DefaultJSONEncoder)
+
+        mongo_client = _connect_mongodb(
+            db_host=properties.get("db_host", "localhost"),
+            db_port=properties.get("db_port", 27017),
+            db_username=properties.get("db_username"),
+            db_password=properties.get("db_password"),
+        )
+        self.mongo_db = mongo_client[properties.get("db_name")]
+
+        if not self._last_edit_date:
+            self.unlock_edit()
+
+    def _check_exposed_type(self, exposed_type):
+        if isinstance(exposed_type, str) and exposed_type not in self.__VALID_STRING_EXPOSED_TYPES:
+            raise InvalidExposedType(
+                f"Invalid string exposed type {exposed_type}. Supported values are {', '.join(self.__VALID_STRING_EXPOSED_TYPES)}"
+            )
+
+    @classmethod
+    def storage_type(cls) -> str:
+        return cls.__STORAGE_TYPE
+
+    @property  # type: ignore
+    @_self_reload(DataNode._MANAGER_NAME)
+    def encoder(self):
+        return self._encoder
+
+    @encoder.setter
+    def encoder(self, encoder: json.JSONEncoder):
+        self.properties[self._ENCODER_KEY] = encoder
+
+    @property  # type: ignore
+    @_self_reload(DataNode._MANAGER_NAME)
+    def decoder(self):
+        return self._decoder
+
+    @decoder.setter
+    def decoder(self, decoder: json.JSONDecoder):
+        self.properties[self._DECODER_KEY] = decoder
+
+    def _read(self):
+        if self.properties[self.__EXPOSED_TYPE_PROPERTY] == self.__EXPOSED_TYPE_PANDAS:
+            return self._read_as_pandas_dataframe()
+        if self.properties[self.__EXPOSED_TYPE_PROPERTY] == self.__EXPOSED_TYPE_NUMPY:
+            return self._read_as_numpy()
+        return self._read_as()
+
+    def _read_as(self):
+        custom_class = self.properties[self.__EXPOSED_TYPE_PROPERTY]
+        cursor = self._read_by_query()
+
+        return list(map(lambda row: custom_class(**row), cursor))
+
+    def _read_as_numpy(self):
+        return self._read_as_pandas_dataframe().to_numpy()
+
+    def _read_as_pandas_dataframe(self, columns: Optional[List[str]] = None):
+        cursor = self._read_by_query()
+        if columns:
+            return pd.DataFrame(list(cursor))[columns]
+        return pd.DataFrame(list(cursor))
+
+    def _read_by_query(self):
+        """Query from MongoDB, exclude the _id field"""
+
+        read_collection = self.mongo_db[self.properties.get("read_collection")]
+        query_result = list(read_collection.find(self.read_query, {"_id": 0}))
+
+        encoded_json = json.dumps(query_result)
```

Comment:
I'm not sure why we need to convert the return from MongoDB to json? Is there a reason for it?

Reply:
This is to create JSON string for the JSONDecoder in `json.load` (line 175) to decode custom Object, or datetime, or Enum from the return of MongoDB.

I'm not sure if there is a better method to decode custom data tyle directly.

Reply:
I believe the user should provide the CustomClass representing the document, and the methods to both convert (encode) from the CustomClass to the MongoDB format, and convert (decode) from the MongoDB format to the CustomClass.

Reply:
Ah I understand what you were trying to do now, the thing that I'm a bit confused is that, if we provide either a pymongo type output, or a json type output, the user will still need to provide an encoding and decoding method right? If yes, then do we need json as the middle-man when we can provide the pymongo output directly? Either way, whether it's JSON or pymongo output, I know for certain we have to document this carefully so the user knows what to expect and provide the CustomClass and its encoder/decoder methods (if any) better :D